File: src/main/java/songscribe/ui/message/LayoutChangeMessage.java
package songscribe.ui.message;

import org.jetbrains.annotations.Nullable;

import songscribe.music.Line;

public class LayoutChangeMessage extends Message {

    public enum ChangeType {
        CONTENT,
        FONT,
        SIZE
    }

    public enum Section {
        TITLE,
        ATTRIBUTION,
        SCORE,
        LYRICS,
        BANGLA_LYRICS,
        TRANSLATION,
        FOOTNOTES
    }

    private final Section section;
    private final ChangeType changeType;
    private final boolean heightChanged;
    @Nullable private final Line line;

    public LayoutChangeMessage(Section section, ChangeType changeType, boolean heightChanged) {
        this(section, changeType, heightChanged, null);
    }

    public LayoutChangeMessage(Section section, ChangeType changeType, boolean heightChanged, @Nullable Line line) {
        this.section = section;
        this.changeType = changeType;
        this.heightChanged = heightChanged;
        this.line = line;
    }

    public static LayoutChangeMessage scoreContent() {
        return scoreContent(null);
    }

    public static LayoutChangeMessage scoreContent(@Nullable Line line) {
        return new LayoutChangeMessage(Section.SCORE, ChangeType.CONTENT, false, line);
    }

    public Section getSection() {
        return section;
    }

    public ChangeType getChangeType() {
        return changeType;
    }

    public boolean isHeightChanged() {
        return heightChanged;
    }

    @Nullable
    public Line getLine() {
        return line;
    }

    @Override
    public String toString() {
        return super.toString() + "(section=" + section + ", changeType=" + changeType
            + ", heightChanged=" + heightChanged + ", line=" + line + ")";
    }
}

File: src/main/java/songscribe/ui/message/LayoutChangeMessage.kt
/*
    SongScribe song notation program
    Copyright (C) Sri Chinmoy Centres International

    This file is part of SongScribe.

    SongScribe is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 3 of the License, or
    (at your option) any later version.

    SongScribe is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

package songscribe.ui.message

import songscribe.music.Line

/**
 * Message posted to MessageCenter when a layout-affecting change occurs.
 *
 * @property section Which section changed
 * @property changeType The type of change (content, font, or size)
 * @property heightChanged Whether the change requires recalculation of sections below
 * @property line The specific line affected, or null if all lines are affected
 */
class LayoutChangeMessage @JvmOverloads constructor(
    val section: Section,
    val changeType: ChangeType,
    val heightChanged: Boolean,
    val line: Line? = null
) : Message() {

    enum class ChangeType {
        CONTENT,
        FONT,
        SIZE
    }

    enum class Section {
        TITLE,
        ATTRIBUTION,
        SCORE,
        LYRICS,
        BANGLA_LYRICS,
        TRANSLATION,
        FOOTNOTES
    }

    override fun toString(): String {
        return "${super.toString()}(section=$section, changeType=$changeType, heightChanged=$heightChanged, line=$line)"
    }
}
